java-checks/src/main/java/org/sonar/java/checks/OutputStreamOverrideWriteCheck.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import org.sonar.check.Rule;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.TypeCriteria;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S4349")
+public class OutputStreamOverrideWriteCheck extends IssuableSubscriptionVisitor {
+
+  private static final MethodMatcher WRITE_BYTES_INT_INT = MethodMatcher.create().typeDefinition(TypeCriteria.anyType()).name("write")
+    .addParameter("byte[]")
+    .addParameter("int")
+    .addParameter("int");
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if(!hasSemantic()) {
+      return;
+    }
+    ClassTree classTree = (ClassTree) tree;
+    Type superType = classTree.symbol().superClass();
+    IdentifierTree className = classTree.simpleName();
+    if(className == null || superType == null || !(superType.is("java.io.OutputStream") || superType.is("java.io.FilterOutputStream"))) {
+      return;
+    }
+    Optional<MethodTree> writeByteIntInt = classTree.members()
+      .stream()
+      .filter(m -> m.is(Tree.Kind.METHOD))
+      .map(m -> (MethodTree) m)
+      .filter(WRITE_BYTES_INT_INT::matches)
+      .findFirst();
+
+    if(!writeByteIntInt.isPresent()) {
```

Comment:
I think it's not enough, because in case of a abstract direct subtypes of `OutputStream`, and not implementing `write(int b)` this may raise a FP. Why asking for implementing `write(byte[], int, int)`, while `write(int)` is not even implemented?

For direct subtypes of `FilterOutputStream` it should be enough, as the method is implemented.

Reply:
Your concern is legit, given the very narrow domain of application of the rule I would suggest to try the naive approach at first and wait for feedback.
Because this would be a FP only if abstract class is not overriding and the subclasses are implementing.

Reply:
Mmh. I don't like the idea of waiting for feedback in such cases. I would prefer the opposite approach and avoid the FP directly, right now. As it's a possible case, even really unlikely. We can predict it, so we can avoid it. Removing any potential noise.